File: Hangman.java
import java.util.ArrayList;
import java.util.Random;
import java.util.Scanner;

public class Hangman {
    private static String constructSecretWord(String string) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < string.length(); i++) {
            if (string.charAt(i) != ' ')
                sb.append('*');
            else
                sb.append(' ');
        }
        return sb.toString();
    }

    private static boolean characterInWord(String word, char character) {
        for (int i = 0; i < word.length(); i++)
            if (word.charAt(i) == character)
                return true;
        return false;
    }

    private static String updateSecretWord(String secretWord, char character, ArrayList<Integer> indexOccurrence) {
        char[] asterics = secretWord.toCharArray();
        for (Integer i: indexOccurrence) {
            if (i == 0) {
                asterics[i] = (char) (character - 32);
            } else {
                if (secretWord.charAt(i - 1) == ' ')
                    asterics[i] = (char) (character - 32);
                else
                    asterics[i] = character;
            }
        }
        return new String(asterics);
    }

    private static ArrayList<Integer> indexOccurrence(String word, char character) {
        ArrayList<Integer> indexOccurrance = new ArrayList<>();
        for (int i = 0; i < word.length(); i++)
            if (word.charAt(i) == character)
                indexOccurrance.add(i);
        return indexOccurrance;
    }

    public static void main(String[] args) {
        Random random = new Random();
        Scanner scanner = new Scanner(System.in);
        String[] words = "Pimpernell,Lyuda,HellFire,Volcano,Conferrence Call,Interfacer,Lady Fist,Sand Hawk,Fibber".split(",");
        String word = words[random.nextInt(words.length)].toLowerCase();
        String secretWord = constructSecretWord(word);
        ArrayList<String> missedLetters = new ArrayList<>();
        int attempts = 7;

        while (attempts != 0) {
            System.out.printf("%s\nMissed letters: %s\n", secretWord, missedLetters.toString());
            System.out.print("\nEnter a character: ");
            String input = scanner.nextLine();
            if (input.length() == 1 && !input.equals("")) {
                char character = input.charAt(0);
                if (characterInWord(word, character)) {
                    ArrayList<Integer> characterIndex = indexOccurrence(word, character);
                    secretWord = updateSecretWord(secretWord, character, characterIndex);
                    if (secretWord.toLowerCase().equals(word)) {
                        System.out.printf("%s\nYou win!\n", secretWord);
                        scanner.close();
                        break;
                    }
                } else {
                    if (!missedLetters.contains(input))
                        missedLetters.add(input);
                    if (--attempts == 0) {
                        System.out.println("You lose!");
                        scanner.close();
                        break;
                    }
                }
            } else {
                System.out.println("Enter one character.");
            }
        }
    }
}

File: Hangman.cs

using System;
using System.Collections.Generic;
using System.Text;

public class Hangman
{
    private static string ConstructSecretWord(string word)
    {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < word.Length; i++) {
            if (word[i] != ' ')
                sb.Append('*');
            else
                sb.Append(' ');
        }
        return sb.ToString();
    }

    private static bool CharacterInWord(string word, char character)
    {
        for (int i = 0; i < word.Length; i++)
            if (word[i] == character)
                return true;
        return false;
    }

    private static List<int> IndexOccurrence(String word, char character)
    {
        List<int> indexOccurrance = new List<int>();
        for (int i = 0; i < word.Length; i++)
            if (word[i] == character)
                indexOccurrance.Add(i);
        return indexOccurrance;
    }

    private static string UpdateSecretWord(String secretWord, char character, List<int> indexOccurrence)
    {
        char[] asterics = secretWord.ToCharArray();
        foreach (var i in indexOccurrence) {
            if (i == 0) {
                asterics[i] = (char) (character - 32); // Uppercase first letter
            } else {
                if (secretWord[i - 1] == ' ')
                    asterics[i] = (char) (character - 32); // Uppercase letter after space
                else
                    asterics[i] = character;
            }
        }
        return new string(asterics);
    }

    public static void Main(string[] args)
    {
        Random random = new Random();
        string[] words = "Pimpernell,Lyuda,HellFire,Volcano,Conferrence Call,Interfacer,Lady Fist,Sand Hawk,Fibber".Split(',');
        string word = words[random.Next(words.Length)].ToLower();
        string secretWord = ConstructSecretWord(word);
        List<string> missedLetters = new List<string>();
        int attempts = 7;

        while (attempts != 0) {
            Console.WriteLine("{0}\nMissed letters: {1}", secretWord, string.Join(", ", missedLetters.ToArray())); // missedLetters.toString()
            Console.Write("\nEnter a character: ");
            string input = Console.ReadLine();
            if (input.Length == 1 && !input.Equals("")) {
                char character = input.ToLower()[0];
                if (CharacterInWord(word, character)) {
                    List<int> characterIndex = IndexOccurrence(word, character);
                    secretWord = UpdateSecretWord(secretWord, character, characterIndex);
                    if (secretWord.ToLower().Equals(word)) {
                        Console.WriteLine("{0}\nYou win!", secretWord);
                        break;
                    }
                }
                else {
                    if (!missedLetters.Contains(input))
                        missedLetters.Add(input);
                    if (--attempts == 0) {
                        Console.WriteLine("You lose. The word was: {0}", word);
                        break;
                    }
                }
            }
            else {
                Console.WriteLine("Enter one character.");
            }
        }
    }
}